packages/runtime/container-runtime/src/blobManager.ts

```diff
@@ -61,40 +63,95 @@ export interface IBlobManagerLoadInfo {
+    private readonly logger: ITelemetryLogger;
+
+    // Map of local (offline/detached) IDs to storage IDs. Contains identity entries
+    // (id -> id) for storage IDs, so all requested IDs should be a key in this map.
+    private readonly redirectTable: Map<string, string | undefined>;
+
+    // Blobs which have not been uploaded or for which we have not yet seen a BlobAttach op round-trip
+    private readonly pendingBlobs: Map<string, PendingBlob> = new Map();
+
+    private readonly retryThrottler = new Throttler(
+        60 * 1000, // 60 sec delay window
+        30 * 1000, // 30 sec max delay
+        // throttling function increases exponentially (0ms, 40ms, 80ms, 160ms, etc)
+        formExponentialFn({ coefficient: 20, initialDelay: 0 }),
+    );
 
     constructor(
         private readonly routeContext: IFluidHandleContext,
         snapshot: IBlobManagerLoadInfo,
         private readonly getStorage: () => IDocumentStorageService,
-        private readonly attachBlobCallback: (blobId: string) => void,
+        private readonly sendBlobAttachOp: (blobId?: string, localId?: string) => void,
         // To be called when a blob node is requested. blobPath is the path of the blob's node in GC's graph. It's
         // of the format `/<BlobManager.basePath>/<blobId>`.
         private readonly gcNodeUpdated: (blobPath: string) => void,
-        private readonly runtime: IContainerRuntime,
-        private readonly logger: ITelemetryLogger,
+        private readonly runtime: IBlobManagerRuntime,
     ) {
-        this.runtime.once("dispose", () => {
-            for (const promise of this.pendingBlobIds.values()) {
-                promise.reject(new Error("runtime disposed while blobAttach op in flight"));
-            }
-        });
-        this.load(snapshot);
+        this.logger = ChildLogger.create(this.runtime.logger, "BlobManager");
+        this.runtime.on("disconnected", () => this.onDisconnected());
+        this.redirectTable = this.load(snapshot);
     }
 
-    private hasBlob(id: string): boolean {
-        return this.blobIds.has(id) || this.detachedBlobIds.has(id);
+    public get pendingOfflineUploadCount(): number {
+        return Array.from(this.pendingBlobs.values())
+            .filter((e) => e.status === PendingBlobStatus.OfflinePendingUpload).length;
+    }
+
+    public get hasPendingOfflineUploads(): boolean {
+        return this.pendingOfflineUploadCount > 0;
+    }
+
+    /**
+     * Upload blobs added while offline. This must be completed before connecting and resubmitting ops.
+     */
+    public async onConnected() {
+        await PerformanceEvent.timedExecAsync(this.logger, {
+                eventName: "BlobUploadOnConnected",
+                count: this.pendingOfflineUploadCount,
+            }, async () => Promise.all(Array.from(this.pendingBlobs.values())
+                    .filter((e) => e.status === PendingBlobStatus.OfflinePendingUpload).map(async (e) => e.uploadP)),
```

Comment:
Nit: I think this is the same filter expression as 'pendingOfflineUploadCount'?

Perhaps 'this.pendingOfflineUploads' would make a good helper?